From 06f0a6a8b93cc28743a444f2644478589e7c6228 Mon Sep 17 00:00:00 2001
From: Benjamin Tissoires <benjamin.tissoires@redhat.com>
Date: Mon, 22 Dec 2014 16:08:35 -0500
Subject: [input] alps: v7 - fix finger counting for > 2 fingers on clickpads

Message-id: <1419264516-15459-4-git-send-email-benjamin.tissoires@redhat.com>
Patchwork-id: 102383
O-Subject: [PATCH RHEL7.1 3/4] Input: alps - v7: fix finger counting for > 2 fingers on clickpads
Bugzilla: 1107819
RH-Acked-by: Hans de Goede <hdegoede@redhat.com>
RH-Acked-by: Aristeu S. Rozanski F. <arozansk@redhat.com>

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1107819
Upstream: will be shipped in v3.20
Brew: https://brewweb.devel.redhat.com/taskinfo?taskID=8420761
Tested by me on a Toshiba Tecra W50-A with a v7 Alps touchpad

commit d27eb7931c98a1ebfc9b2fcc48939846bcbfc804
(in the input maintainer's tree:
https://git.kernel.org/cgit/linux/kernel/git/dtor/input.git)
Author: Hans de Goede <hdegoede@redhat.com>
Date:   Thu Dec 18 09:55:14 2014 -0800

    Input: alps - v7: fix finger counting for > 2 fingers on clickpads

    Protocol v7 uses the middle / right button bits on clickpads to communicate
    "location" information of a 3th touch (and possible 4th) touch on
    clickpads.

    Specifically when 3 touches are down, if one of the 3 touches is in the
    left / right button area, this will get reported in the middle / right
    button bits and the touchpad will still send a TWO type packet rather then
    a MULTI type packet, so when this happens we must add the finger reported
    in the button area to the finger count.

    Likewise we must also add fingers reported this way to the finger count
    when we get MULTI packets.

    BugLink: https://bugs.freedesktop.org/show_bug.cgi?id=86338
    Cc: stable@vger.kernel.org # 3.17
    Signed-off-by: Hans de Goede <hdegoede@redhat.com>
    Tested-by: Benjamin Tissoires <benjamin.tissoires@redhat.com>
    Signed-off-by: Dmitry Torokhov <dmitry.torokhov@gmail.com>

Signed-off-by: Benjamin Tissoires <benjamin.tissoires@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/input/mouse/alps.c b/drivers/input/mouse/alps.c
index 4623321..c177b69 100644
--- a/drivers/input/mouse/alps.c
+++ b/drivers/input/mouse/alps.c
@@ -933,6 +933,7 @@ static int alps_decode_packet_v7(struct alps_fields *f,
       unsigned char *p,
       struct psmouse *psmouse)
 {
+ struct alps_data *priv = psmouse->private;
  unsigned char pkt_id;
 
  pkt_id = alps_get_packet_id_v7(p);
@@ -963,15 +964,22 @@ static int alps_decode_packet_v7(struct alps_fields *f,
 
  alps_get_finger_coordinate_v7(f->mt, p, pkt_id);
 
- f->left = (p[0] & 0x80) >> 7;
- f->right = (p[0] & 0x20) >> 5;
- f->middle = (p[0] & 0x10) >> 4;
-
  if (pkt_id == V7_PACKET_ID_TWO)
   f->fingers = alps_get_mt_count(f->mt);
  else /* pkt_id == V7_PACKET_ID_MULTI */
   f->fingers = 3 + (p[5] & 0x03);
 
+ f->left = (p[0] & 0x80) >> 7;
+ if (priv->flags & ALPS_BUTTONPAD) {
+  if (p[0] & 0x20)
+   f->fingers++;
+  if (p[0] & 0x10)
+   f->fingers++;
+ } else {
+  f->right = (p[0] & 0x20) >> 5;
+  f->middle = (p[0] & 0x10) >> 4;
+ }
+
  /* Sometimes a single touch is reported in mt[1] rather then mt[0] */
  if (f->fingers == 1 && f->mt[0].x == 0 && f->mt[0].y == 0) {
   f->mt[0].x = f->mt[1].x;
-- 
1.7.1